### ISSUE TITLE
How to enable IntelliSense for algorithms?

### ISSUE THREAD

**Issue by kongehund** [fbertram/TuringTrader#19]
When choosing "Edit -> Algorithm Source..." to edit an algorithm, IntelliSense doesn't work because the files are not part of the source solution. (i.e. the Demo scripts)

It's impractical to not have autocomplete or not to be able to go to a method's definition when editing scripts.

What's the recommended way to enable IntelliSense for algorithm scripts?

**Comment by fbertram** [comment]
Hello Kongehund,

TuringTrader has two different methods of loading algorithms:
(1) loading them from C# source code, and compiling them in memory
(2) loading them from a DLL

While method (1) is quick and convenient, it also has severe limitations.  One of them is, that, because the C# file is not part of a solution, IntelliSense is not available. Other restrictions include that there is a one-to-one relationship between C# files and algorithms, making it impossible to split the code across multiple files, or to implement multiple algorithms in a single file.

Therefore, I highly recommend method (2) for more complex algorithms. By creating a DLL, all C# features are available to the developer. The Books&Pubs project (see https://github.com/fbertram/TuringTrader/tree/develop/BooksAndPubsV2), included with the distribution, shows how to do this.

In summary, this is **not a bug**, and there are **no plans to improve** the developer experience for method (1).

Thank you,
Best regards

Felix

**Comment by kongehund** [comment]
Hi Felix,

Thank you for the response.

You mention that the Books&Pubs project shows how to create DLLs. However, I am unable to find the instructions for how to do this, since the project has no readme, and the [docs article](https://www.turingtrader.org/help/articles/qsg/CreateDlls.html) has not been written yet. 

I apologize if you have already written instructions elsewhere, but are you able to link to or summarize the key steps to create DLLs and have them work in TuringTrader? 

I should mention that I have already experimented with creating DLLs (by building BooksAndPubsV2) and these are my observations, which have caused me a bit of confusion as to how TuringTrader works with and locates DLLs:
- When I open `Documents\TuringTrader\BooksAndPubsV2\BooksAndPubsV2.sln` in Visual Studio and right click the project in the solution tree and click Build, my changes are applied and the binary version of TuringTrader successfully runs my changes.
- When I compile TuringTrader from source, there are no BooksAndPubs algorithms available, whereas in binary they are available under "Showcase Algorithms (Legacy)" and "Showcase Algorithms (V2)". The .cs files, however, (Demo files and custom .cs files in `Documents\TuringTrader\Algorithms\` are avilable in both binary distribution and compiled version.
- When I put the DLL files of BooksAndPubsV2 inside the Algorithms folder, the algorithm does not show up in the compiled source version of TuringTrader.

I imagine that creating DLLs may be as simple as creating a class library project in Visual Studio, getting the .cs files into the solution and clicking build, but if I'm correct about this, I don't know how to make TuringTrader locate the DLL.

I don't know if you receive notifications about closed issues, so I will tag you @fbertram 

Thanks in advance,
Kongehund

**Comment by fbertram** [comment]
Hello Kongehund,

A TuringTrader DLL is basically a class library project, plus the following minor aspects:

- The DLL should link against the installed TuringTrader binary at /Program Files/TuringTrader/bin
- We need an additional build step to copy the DLL output into the ~/Documents/TuringTrader/Algorithms folder

The easiest way to get started might be to simply copy the Books & Pubs project.

You are correct that the Books & Pubs projects won't load when launching TuringTrader from MSVC. You can fix this by linking the DLLs against the project (instead of the simulator in /Program Files/TuringTrader/bin). The project is set up to work well for users of the binary distribution; it is assumed that developers running it in MSVC know their way around.

Best regards, Felix

**Comment by kongehund** [comment]
Thank you for the explanation!
